contrib/misc/statefile-parser/statefile.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python3
+
+#   BAREOS® - Backup Archiving REcovery Open Sourced
+#
+#   Copyright (C) 2024-2024 Bareos GmbH & Co. KG
+#
+#   This program is Free Software; you can redistribute it and/or
+#   modify it under the terms of version three of the GNU Affero General Public
+#   License as published by the Free Software Foundation and included
+#   in the file LICENSE.
+#
+#   This program is distributed in the hope that it will be useful, but
+#   WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+#   Affero General Public License for more details.
+#
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program; if not, write to the Free Software
+#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+#   02110-1301, USA.
+
+"""This module parses a Bareos filedaemon statefile.
+
+Passing the raw data of the statefile to parse_statefile() will return a list
+of JobResult objects. A code sample can be found is at the end of this file.
+
+This module will probably not work for files generated with the windows version
+of Bareos, as structs are packed differently.
+"""
+
+from glob import glob
+from struct import unpack, calcsize
+from dataclasses import dataclass
+from datetime import datetime
+
+HEADER_FMT = "".join(
+    (
+        "=",  # native byte-order with standard sizes
+        "14s2x",  # magic
+        "i4x",  # version
+        "Q",  # last job offset
+        "Q",  # end of list
+    )
+)
+HEADER_LEN = calcsize(HEADER_FMT)
+
+NUM_FMT = "=I"  # number of result records
+NUM_FMT_LEN = calcsize(NUM_FMT)
+
+RESULT_FMT = "".join(
+    (
+        "=",  # native byte-order with standard sizes
+        "16x",  # initial padding
+        "i",  # errors
+        "i",  # job type
+        "i",  # job status
+        "i",  # job level
+        "I",  # job id
+        "I",  # volume session id
+        "I",  # volume session time
+        "I",  # job files
+        "Q",  # job bytes
+        "Q",  # job start time
+        "Q",  # job end time
```

Comment:
Some useless trivia on the usability of `time_t`: GNU libc guarantees that `time_t` is signed and negative values are handled correctly, even though they're undefined as of POSIX.
POSIX only requires that `time_t` is "an integer type" that represents seconds since epoch, so a system using `int8_t` would still be compliant, even though it couldn't represent a time after 1970-01-01T00:02:07Z.
Last but not least ISO C defines `time_t` as integer or float that you can pass to library functions that accept a `time_t`.

But yes, in our case `"q"` is more correct than `"Q"` :smiley: